tracing/span.go

```diff
@@ -209,192 +163,463 @@ func (s *Span) ToZipkinSpan() ZipkinSpan {
+type baseSpan struct {
+	trace *trace
+	hooks []SpanHook
+}
+
+// Name returns the name of the Span.
+func (s baseSpan) Name() string {
+	return s.trace.name
+}
+
+// SpanType returns the SpanType of the Span.
+func (s baseSpan) SpanType() SpanType {
+	return SpanTypeUnknown
+}
+
+// Hooks returns all SpanHooks registered to the Span.
+func (s baseSpan) Hooks() []SpanHook {
+	return s.hooks
+}
+
+// AddHooks adds hooks into the Span.
+//
+// It is recommended that you only call AddHooks on a Span within an
+// OnCreate Hook so the Span is set up with all of it's Hooks as a part of
+// it's creation.
+func (s *baseSpan) AddHooks(hooks ...SpanHook) {
+	s.hooks = append(s.hooks, hooks...)
+}
+
+// SetDebug sets or unsets the debug flag of this Span.
+func (s *baseSpan) SetDebug(v bool) {
+	s.trace.setDebug(v)
+}
 
-	for _, hook := range s.hooks {
-		if err := hook.OnAddCounter(s, key, delta); err != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnAddCounter hook error: " + err.Error())
+// Stop calls span.PreStop, calls all registered OnStop Hooks, serializes
+// the Span and sends the serialized Span to a back-end that records the
+// Span.
+func (s *baseSpan) SetTag(key string, value interface{}) {
+	s.trace.setTag(key, value)
+
+	for _, hook := range s.Hooks() {
+		if err := hook.OnSetTag(s, key, value); err != nil {
+			s.LogError("OnStart hook error: ", err)
 		}
 	}
 }
 
-// SetTag sets a binary tag annotation.
-func (s *Span) SetTag(key string, value interface{}) {
-	s.tags[key] = value
+// AddCounter adds delta to a counter annotation and calls all OnAddCounter
+// Hooks registered to the Span.
+func (s *baseSpan) AddCounter(key string, delta float64) {
+	s.trace.addCounter(key, delta)
 
-	for _, hook := range s.hooks {
-		if err := hook.OnSetTag(s, key, value); err != nil && s.tracer.Logger != nil {
-			s.tracer.Logger("OnSetTag hook error: " + err.Error())
+	for _, hook := range s.Hooks() {
+		if err := hook.OnAddCounter(s, key, delta); err != nil {
+			s.LogError("OnStart hook error: ", err)
 		}
 	}
 }
 
-// SetDebug sets or unsets the debug flag of this span.
-func (s *Span) SetDebug(v bool) {
-	if v {
-		s.flags |= FlagMaskDebug
-	} else {
-		s.flags &= ^FlagMaskDebug
+// PreStop is where you can place custom logic for changes you want to make
+// to the Span once it has ended but before it is stopped and published.
+//
+// PreStop does not need to be called directly, instead you should call
+// span.Stop() which combines stopping the Span, calling all OnStop Hooks
+// registered to the Span, serializing the Span, and sending the serialized
+// Span to a back-end that records the Span.
+func (s *baseSpan) PreStop(err error) {
+	return
+}
+
+// Stop calls span.PreStop, calls all registered OnStop Hooks, serializes
+// the Span and sends the serialized Span to a back-end that records the
+// Span.
+func (s *baseSpan) Stop(ctx context.Context, err error) error {
+	s.PreStop(err)
```

Comment:
if you expect this calls PreStop function from the concert type embedding baseSpan (for example, ServerSpan.PreStop), I don't think this works.

Reply:
:doh:

Reply:
I think having a global function is the only way to do that.